webviz_subsurface/plugins/_group_tree/group_tree_data.py

```diff
@@ -0,0 +1,478 @@
+from typing import List, Dict, Tuple, Any
+import logging
+
+import pandas as pd
+import numpy as np
+
+from webviz_config.common_cache import CACHE
+
+
+class GroupTreeData:
+    """This class holds the summary and gruptree datasets and functionality
+    to combine the data and calculate the GroupTree component input dataset.
+    """
+
+    def __init__(self, smry: pd.DataFrame, gruptree: pd.DataFrame):
+        self.smry = smry
+        self.gruptree = gruptree
+        self.ensembles = gruptree["ENSEMBLE"].unique()
+        self.wells = gruptree[gruptree["KEYWORD"] == "WELSPECS"]["CHILD"].unique()
+
+        # Check that all field rate summary vectors exist
+        self.check_that_sumvecs_exists(["FOPR", "FGPR", "FWPR", "FWIR", "FGIR"])
+
+        # Check that WSTAT exists for all wells
+        self.check_that_sumvecs_exists([f"WSTAT:{well}" for well in self.wells])
+
+        # Check if the ensembles have waterinj and/or gasinj
+        self.has_waterinj, self.has_gasinj = {}, {}
+        for ensemble in self.ensembles:
+            smry_ens = self.smry[self.smry["ENSEMBLE"] == ensemble]
+            self.has_waterinj[ensemble] = smry_ens["FWIR"].sum() > 0
+            self.has_gasinj[ensemble] = smry_ens["FGIR"].sum() > 0
+
+        # Add nodetypes IS_PROD, IS_INJ and IS_OTHER to gruptree
+        self.gruptree = add_nodetype(self.gruptree, self.smry)
+
+        # Add edge label
+        self.gruptree["EDGE_LABEL"] = self.gruptree.apply(get_edge_label, axis=1)
+
+        # Get summary data with metadata (ensemble, nodename, datatype, edge_or_node)
+        self.sumvecs = self.get_sumvecs_with_metadata()
+
+        # Check that all summary vectors exist
+        self.check_that_sumvecs_exists(list(self.sumvecs["SUMVEC"]))
+
+    @CACHE.memoize(timeout=CACHE.TIMEOUT)
+    def create_grouptree_dataset(
+        self,
+        ensemble: str,
+        tree_mode: str,
+        real: int,
+        prod_inj_other: List[str],
+    ) -> Tuple[List[Dict[Any, Any]], List[Dict[str, str]], List[Dict[str, str]]]:
+        """This method is called when an event is triggered to create a new dataset
+        to the GroupTree plugin. First there is a lot of filtering of the smry and
+        grouptree data, before the filtered data is sent to the function that is
+        actually creating the dataset.
+
+        Returns the group tree data and two lists with dropdown options for what
+        to display on the edges and nodes.
+
+        A sample data set can be found here:
+        https://github.com/equinor/webviz-subsurface-components/blob/master/react/ \
```

Comment:
leave on one line and add `'# noqa` after the final `"""` will make copy and paste easier ;)